### ISSUE TITLE
On import and generation of .olean files

### ISSUE THREAD

**Issue by bbentzen** [leanprover/lean3#986]
So I have a folder with a `.project` file and two `.lean` files - `foo.lean` and `main.lean`. I need to load definitions and theorems from the former to the latter.
But writing `import .foo` gives me two errors:
1. `failed to compile ...\foo.lean`
2. `invalid import, unknown module 'foo'.`

I suspect this is because I don't have a `foo.olean` file. When I try to compile it on emacs (assuming that doing this will generate my `foo.olean` file), it keeps giving me the message:

```
-*- mode: compilation; default-directory: "c:......../" -*-
Compilation started at Mon Feb  1 19:49:36

make -k 
'make' is not recognized as an internal
or external command, operable program or batch file

Compilation exited abnormally with code 1 at Mon Feb  1 19:49:36
```

I've followed all the instructions of [this stack overflow answer](http://stackoverflow.com/a/23734705), (I've download  'make-3.81.exe' and changed my PATH environment variable) but the problem persists. If I type'C:\Program Files (x86)\GnuWin32\bin\make.exe' directly in the compile command I get the error:

`make: *** No targets specified and no makefile found.  Stop.`

instead.

Is this the only way to load theorems from another file in Lean / generating `.olean` files?

By the way, I'm using Windows 7.

**Comment by soonhokong** [comment]
@bbentzen: could you try the following and let me know the output you get?
1. Make a new directory, put the following three files:
   
   .project
   
   ```
   # Lean project file
   
   # Include all .lean files under this directory
   + *.lean
   
   # Exclude flycheck generated temp files
   - flycheck*.lean
   
   # Exclude emacs temp files
   - .#*.lean
   ```
   
   foo.lean
   
   ``` lean
   import standard
   namespace mine
   definition id (A : Type) (x : A) := x
   end mine
   ```
   
   main.lean
   
   ``` lean
   import .foo
   check mine.id
   ```
2. Run `linja` in that directory (i.e. run `c:/lean/bin/linja`)

**Comment by bbentzen** [comment]
This gives me

```
import -- failed to compile \Users\Bruno\...\foo.lean
.foo -- invalid import, unknown module 'foo'

check mine.id -- unknown identifier 'mine.id'
```

I am not sure what you mean by "running" linja, since it is not a `.exe` file (I am using Windows). Do you mean compiling it with emacs? If this is so, I get the same error

```
make -k 
'make' is not recognized as an internal
or external command, operable program or batch file

Compilation exited abnormally with code 1 at Mon Feb  1 19:49:36
```

as before.

**Comment by soonhokong** [comment]
> I am not sure what you mean by "running" linja, since it is not a .exe file (I am using Windows). 

`linja` is a python script which triggers a project build (following its dependencies). It should be located at the same directory where `lean.exe` is located (i.e. under `c:\lean\bin`).

Can you find it and run it at the directory where you have the other files -- `.project`, `foo.lean`, and `main.lean`?

> Do you mean compiling it with emacs? If this is so, I get the same error.

No, I wanted to check the problem without Emacs getting involved.

**Comment by bbentzen** [comment]
(I think) I got it. I copied linja to the same directory as the other files (and I had to copy the other `lean/bin` files too) and then: 

`ninja: error: expected depfile 'Users\Bruno\...\foo.d' to mention 'Users\Bruno\...\foo.olean' got 'Users\Bruno\...\foo.olean'`

This is what I get when I try to run it.

**Comment by soonhokong** [comment]
@bbentzen, could you send me your `build.ninja` file? My email is `soonhok at cs dot cmu dot edu`.

**Comment by soonhokong** [comment]
@bbentzen, thanks for sending the file. I think there is a problem in handling spaces inside of a directory name (i.e. `My education`) in Windows platform. Could you move the files to another directory whose fullpath doesn't contain any space? If this works, we can localize the problem.

I'll take care of the space problem separately, of course.

**Comment by bbentzen** [comment]
I confirm that this works: `import .foo` runs normally now, but only after I moved all the project files to another directory which does not contain any blank spaces in its path.

**Comment by bbentzen** [comment]
For the record, I tried to reproduce the problem using `.hlean` files and I found out that I get the `invalid import` error again - even in a directory with path without blank spaces. Running `linja` in this new directory gives me the output `ninja: no work to do`.

I guess the cause of the problem is that the automatically generated `.project` is only applicable to `.lean` files?

```
# Lean project file

# Include all .lean files under this directory
+ *.lean

# Exclude flycheck generated temp files
- flycheck*.lean

# Exclude emacs temp files
- .#*.lean
```

If this is the case, what should I do to enable `import` in the HoTT library?

**Comment by javra** [comment]
Exchanging all occuences of `.lean` by `.hlean` works. I guess the emacs-plugin should actually have an option to choose...

**Comment by bbentzen** [comment]
Yes, it does. But as far as I can tell, the emacs mode only has the `create a new project` option that creates directly the `.project` file described above, at least in my version of lean (lean-0.2.0-windows). It would be very nice to have an option to chose, though.

**Comment by soonhokong** [comment]
@bbentzen, could you test my patch?

https://github.com/soonhokong/lean/commit/1deb7ac45f6b6b1ddbbacdc35cb46f2281e0126a

**Comment by soonhokong** [comment]
For the `.lean`/`.hlean` issue, I'll update `lean-create-project` elisp function to use the extension of the current file.

**Comment by bbentzen** [comment]
@soonhokong, this solves the problem.

**Comment by soonhokong** [comment]
Closed by d8fb6f5.

**Comment by soonhokong** [comment]
d8fb6f5 causes a problem on Linux. Let me roll back and commit another patch.

**Comment by bbentzen** [comment]
By the way, I realized that if I want to add any new definition to `foo.hlean` posteriorly, i.e. after creating the `.project`, `foo.hlean` and `main.hlean` files with `import .foo` in `main.hlean`, I get an `unknown identifier 'new-def-name'` error when I type `check new-def-name`, while the other old definitions of `foo` type-check normally - suggesting that only the definitions that were present at the time of the generation of the `foo.olean` are recognized by the project, as if the `.olean` file generated isn't being "updated" somehow.

Is there any work-around for this problem?

**Comment by leodemoura** [comment]
The latest version does not use `linja` anymore.
